C++ compile error redefinition of struct @ 4f1f487ae

```
g++ -O0  -I.  -MMD -I.../verilator/test_regress/../include -I.../verilator/test_regress/../include/vltstd -DVM_COVERAGE=0 -DVM_SC=0 -DVM_TRACE=0 -DVM_TRACE_FST=0 -DVM_TRACE_VCD=0 -faligned-new -fcf-protection=none -Wno-bool-operation -Wno-sign-compare -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-unused-variable -Wno-shadow       -DVERILATOR=1 -DVL_DEBUG=1 -DTEST_OBJ_DIR=obj_vlt/t_class_typedef -DVM_PREFIX=Vt_class_typedef -DVM_PREFIX_INCLUDE="<Vt_class_typedef.h>" -DVM_PREFIX_ROOT_INCLUDE="<Vt_class_typedef___024root.h>" -DT_CLASS_TYPEDEF   -c -o Vt_class_typedef__ALL.o Vt_class_typedef__ALL.cpp
In file included from Vt_class_typedef___024root.h:10,
                 from Vt_class_typedef__Syms.h:17,
                 from Vt_class_typedef.cpp:5,
                 from Vt_class_typedef__ALL.cpp:3:
Vt_class_typedef___024unit__03a__03aB__W3__Vclpkg.h:15:8: error: redefinition of ‘struct Vt_class_typedef_t_struct__struct3’
 struct Vt_class_typedef_t_struct__struct3 {
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from Vt_class_typedef___024root.h:9,
                 from Vt_class_typedef__Syms.h:17,
                 from Vt_class_typedef.cpp:5,
                 from Vt_class_typedef__ALL.cpp:3:
Vt_class_typedef___024unit__03a__03aB__Vclpkg.h:15:8: note: previous definition of ‘struct Vt_class_typedef_t_struct__struct3’
 struct Vt_class_typedef_t_struct__struct3 {
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from Vt_class_typedef__ALL.cpp:13:
Vt_class_typedef___024unit__03a__03aB__W3__Vclpkg__DepSet_hf3df3bd9__0.cpp:10:13: error: redefinition of ‘std::__cxx11::string VL_TO_STRING(const Vt_class_typedef_t_struct__struct3&)’
 std::string VL_TO_STRING(const Vt_class_typedef_t_struct__struct3& obj) {
             ^~~~~~~~~~~~
In file included from Vt_class_typedef__ALL.cpp:11:
Vt_class_typedef___024unit__03a__03aB__Vclpkg__DepSet_hadfb9064__0.cpp:10:13: note: ‘std::__cxx11::string VL_TO_STRING(const Vt_class_typedef_t_struct__struct3&)’ previously defined here
 std::string VL_TO_STRING(const Vt_class_typedef_t_struct__struct3& obj) {
             ^~~~~~~~~~~~
```

Can you attach an example that shows the issue?  (Must be openly licensed, ideally in test_regress format.)

```
class uvm_resource_types;
   typedef int rsrc_q_t;
endclass

class uvm_resource_pool;
   uvm_resource_types::rsrc_q_t rtab [string];
endclass

class B#(parameter WIDTH = 2);
    typedef logic [WIDTH-1:0] t_vector;
    typedef struct {
        t_vector m0;
    } t_struct;
endclass

module t_io_class_typedef   #(parameter WIDTH = 2) (
   // Inputs
   input  B#(WIDTH)::t_struct in_struct,
   // Outputs
   output B#(WIDTH)::t_vector out_vector
   );

   assign out_vector = in_struct.m0 + 2'd1;

endmodule

virtual class C#(parameter type T = logic, parameter SIZE = 1);
    typedef logic [SIZE-1:0] t_vector;
    typedef T t_array [SIZE-1:0];
    typedef struct {
        t_vector m0 [2*SIZE-1:0];
        t_array m1;
    } t_struct;
endclass

module t (/*AUTOARG*/
   // Inputs
   clk
   );
   input clk;

   B#()::t_struct b_struct1;
   B#()::t_vector b_vector1;

   B#(3)::t_struct b_struct2;
   B#(3)::t_vector b_vector2;

   assign b_struct1.m0 = 2'd1;
   assign b_struct2.m0 = 3'd2;

   t_io_class_typedef i_t_io_class_typedef1(.in_struct (b_struct1),
					    .out_vector(b_vector1));

   t_io_class_typedef #(3) i_t_io_class_typedef2 (.in_struct (b_struct2),
					          .out_vector(b_vector2));

   initial begin
      uvm_resource_pool pool = new;
      typedef logic [7:0] t_t0;
      C#(t_t0,3)::t_vector v0;
      C#(t_t0,3)::t_array a0;
      C#(bit,4)::t_struct s0;

      pool.rtab["a"] = 1;
      if ($bits(pool.rtab["a"]) != 32) $stop;

      if ($bits(v0) != 3) $stop;
      if ($size(a0) != 3) $stop;
      if ($bits(a0[0]) != 8) $stop;
      if ($size(s0.m0) != 8) $stop;
      if ($size(s0.m1) != 4) $stop;
      if ($bits(s0.m1[2]) != 1) $stop;
      if ($bits(b_struct1.m0) != 2) $stop;
      if ($bits(b_vector1) != 2) $stop;
      if ($bits(b_struct2.m0) != 3) $stop;
      if ($bits(b_vector2) != 3) $stop;
   end

   always @ (posedge clk) begin
      if (b_vector1 != 2'd2) $stop;
      if (b_vector2 != 3'd3) $stop;

      $write("*-* All Finished *-*\n");
      $finish;
   end

endmodule
```

What 'verilator --version' are you using?  Did you try it with the git master version?

**Master @ https://github.com/verilator/verilator/commit/4f1f487ae4ea0811883e78d287941a4d8abbad19**

What OS and distribution are you using?

```
uname -a
Linux loran 4.18.0-492.el8.x86_64 #1 SMP Tue May 9 17:56:55 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

May we assist you in trying to fix this in Verilator yourself?

**Not sure**

--- a/src/V3Class.cpp
+++ b/src/V3Class.cpp
@@ -183,7 +183,8 @@ private:
         // Give struct a pointer to its package and a final name
         dtypep->editCountInc();
         dtypep->classOrPackagep(m_classPackagep ? m_classPackagep : m_modp);
-        dtypep->name(m_names.get(dtypep->name() + (VN_IS(dtypep, UnionDType) ? "__union" : "__struct")));
+        dtypep->name(
+            m_names.get(dtypep->name() + (VN_IS(dtypep, UnionDType) ? "__union" : "__struct")));
 
         for (const AstMemberDType* itemp = dtypep->membersp(); itemp;
              itemp = VN_AS(itemp->nextp(), MemberDType)) {

--- a/src/V3Class.cpp
+++ b/src/V3Class.cpp
@@ -27,6 +27,7 @@
 
 #include "V3Ast.h"
 #include "V3Global.h"
+#include "V3UniqueNames.h"
 
 VL_DEFINE_DEBUG_FUNCTIONS;
 
@@ -41,6 +42,7 @@ private:
 
     // MEMBERS
     string m_prefix;  // String prefix to add to name based on hier
+    V3UniqueNames m_names;  // For unique naming of structs and unions
     AstNodeModule* m_modp = nullptr;  // Current module
     AstNodeModule* m_classPackagep = nullptr;  // Package moving into
     const AstScope* m_classScopep = nullptr;  // Package moving scopes into
@@ -181,8 +183,7 @@ private:
         // Give struct a pointer to its package and a final name
         dtypep->editCountInc();
         dtypep->classOrPackagep(m_classPackagep ? m_classPackagep : m_modp);
-        dtypep->name(dtypep->name() + (VN_IS(dtypep, UnionDType) ? "__union" : "__struct")
-                     + cvtToStr(dtypep->uniqueNum()));
+        dtypep->name(m_names.get(dtypep->name() + (VN_IS(dtypep, UnionDType) ? "__union" : "__struct")));
 
         for (const AstMemberDType* itemp = dtypep->membersp(); itemp;
              itemp = VN_AS(itemp->nextp(), MemberDType)) {

--- a/test_regress/t/t_unpacked_struct_redef.pl
+++ b/test_regress/t/t_unpacked_struct_redef.pl
@@ -1,0 +1,21 @@
+#!/usr/bin/env perl
+if (!$::Driver) { use FindBin; exec("$FindBin::Bin/bootstrap.pl", @ARGV, $0); die; }
+# DESCRIPTION: Verilator: Verilog Test driver/expect definition
+#
+# Copyright 2023 by Wilson Snyder. This program is free software; you
+# can redistribute it and/or modify it under the terms of either the GNU
+# Lesser General Public License Version 3 or the Perl Artistic License
+# Version 2.0.
+# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+
+scenarios(simulator => 1);
+
+compile(
+    );
+
+execute(
+    check_finished => 1,
+    );
+
+ok(1);
+1;

--- a/test_regress/t/t_unpacked_struct_redef.v
+++ b/test_regress/t/t_unpacked_struct_redef.v
@@ -1,0 +1,26 @@
+// DESCRIPTION: Verilator: Verilog Test module
+//
+// This file ONLY is placed under the Creative Commons Public Domain, for
+// any use, without warranty, 2023 by Antmicro Ltd.
+// SPDX-License-Identifier: CC0-1.0
+
+class Class#(parameter WIDTH);
+    typedef logic [WIDTH-1:0] word;
+    typedef struct {
+        word w;
+    } Struct;
+endclass
+
+module t;
+   Class#(1)::Struct s1;
+   Class#(1)::Struct s2;
+   Class#(2)::Struct s3;
+
+   initial begin
+      $display("%p", s1);
+      $display("%p", s2);
+      $display("%p", s3);
+      $write("*-* All Finished *-*\n");
+      $finish;
+   end
+endmodule

